File: backend/src/handlers/sync.ts
/**
 * Lambda handler: POST /sync
 *
 * Fetches all published pages from Notion, converts them to HTML,
 * stores HTML in S3, and upserts metadata in DynamoDB.
 */
import type { APIGatewayProxyHandler } from "aws-lambda";
import { marked } from "marked";
import {
  fetchNotionPages,
  pageToMarkdown,
  pageToMetadata,
} from "../lib/notion.js";
import {
  putArticle,
  getArticleByNotionId,
} from "../lib/dynamodb.js";
import { putArticleHtml } from "../lib/s3.js";
import { ok, serverError, noContent } from "../lib/response.js";
import type { ArticleRecord, SyncResult } from "@notohub/shared";

export const handler: APIGatewayProxyHandler = async (event) => {
  if (event.httpMethod === "OPTIONS") return noContent();

  const syncedAt = new Date().toISOString();
  const errors: string[] = [];
  let processed = 0;
  let updated = 0;

  try {
    const pages = await fetchNotionPages();
    processed = pages.length;

    for (const page of pages) {
      try {
        // Check if this article already exists
        const existing = await getArticleByNotionId(page.id);
        const existingId = existing?.id;

        // Skip if Notion hasn't changed since last sync
        if (
          existing &&
          existing.notionLastEditedAt === page.last_edited_time
        ) {
          continue;
        }

        // Convert Notion page body → Markdown → HTML
        const markdown = await pageToMarkdown(page.id);
        const html = await marked.parse(markdown);

        // Build metadata (may reuse existing id)
        const metadata = pageToMetadata(page, "", existingId);

        // Store HTML in S3
        const s3Key = await putArticleHtml(metadata.id, html);

        // Build DynamoDB record
        const record: ArticleRecord = {
          PK: `ARTICLE#${metadata.id}`,
          SK: "ARTICLE",
          GSI1PK: `SLUG#${metadata.slug}`,
          GSI1SK: "ARTICLE",
          id: metadata.id,
          notionPageId: page.id,
          title: metadata.title,
          slug: metadata.slug,
          excerpt: metadata.excerpt,
          coverImageUrl: metadata.coverImageUrl,
          s3Key,
          tagsJson: JSON.stringify(metadata.tags),
          authorJson: JSON.stringify(metadata.author),
          publishedAt: metadata.publishedAt,
          updatedAt: new Date().toISOString(),
          notionLastEditedAt: page.last_edited_time,
        };

        await putArticle(record);
        updated++;
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        errors.push(`Page ${page.id}: ${msg}`);
        console.error(`[SYNC] Error processing page ${page.id}:`, err);
      }
    }

    const result: SyncResult = {
      status: errors.length === 0 ? "succeeded" : "failed",
      syncedAt,
      articlesProcessed: processed,
      articlesUpdated: updated,
      errors,
    };

    return ok(result);
  } catch (err) {
    return serverError(err);
  }
};

File: backend/handlers/sync.py
import json
from datetime import datetime, timezone

from lib.notion import fetch_notion_pages, page_to_metadata, page_to_html
from lib.dynamodb import put_article, get_article_by_notion_id
from lib.s3 import put_article_html
from lib.response import ok, server_error


def handle_sync(event: dict) -> dict:
    """
    POST /sync — fetch all Published Notion pages, render to HTML,
    store in S3, and upsert metadata in DynamoDB.
    """
    synced_at = datetime.now(timezone.utc).isoformat()
    errors: list[str] = []
    processed = 0
    updated = 0

    try:
        pages = fetch_notion_pages()
        processed = len(pages)

        for page in pages:
            page_id = page["id"]
            try:
                existing = get_article_by_notion_id(page_id)
                existing_id = existing["id"] if existing else None

                # Skip pages that haven't changed since last sync
                if existing and existing.get("notionLastEditedAt") == page["last_edited_time"]:
                    continue

                metadata = page_to_metadata(page, existing_id=existing_id)

                # Notion blocks → styled HTML (via notion_to_html renderer)
                author_name = metadata["author"]["name"]
                html = page_to_html(page_id, author_name)

                # Upload to s3://<bucket>/{author_name}/{slug}/index.html
                s3_key = put_article_html(author_name, metadata["slug"], html)

                record = {
                    "PK": f"ARTICLE#{metadata['id']}",
                    "SK": "ARTICLE",
                    "GSI1PK": f"SLUG#{metadata['slug']}",
                    "GSI1SK": "ARTICLE",
                    "id": metadata["id"],
                    "notionPageId": page_id,
                    "title": metadata["title"],
                    "slug": metadata["slug"],
                    "excerpt": metadata["excerpt"],
                    "coverImageUrl": metadata.get("coverImageUrl"),
                    "s3Key": s3_key,
                    "tagsJson": json.dumps(metadata["tags"]),
                    "authorJson": json.dumps(metadata["author"]),
                    "publishedAt": metadata["publishedAt"],
                    "updatedAt": datetime.now(timezone.utc).isoformat(),
                    "notionLastEditedAt": page["last_edited_time"],
                }

                put_article(record)
                updated += 1

            except Exception as exc:
                msg = f"Page {page_id}: {exc}"
                errors.append(msg)
                print(f"[SYNC] Error — {msg}", flush=True)

        return ok({
            "status": "succeeded" if not errors else "failed",
            "syncedAt": synced_at,
            "articlesProcessed": processed,
            "articlesUpdated": updated,
            "errors": errors,
        })

    except Exception as exc:
        return server_error(exc)
